crates/bevy_text/src/pipeline.rs

```diff
@@ -1,52 +1,147 @@
-use crate::{
-    compute_text_bounds, error::TextError, glyph_brush::GlyphBrush, scale_value, BreakLineOn, Font,
-    FontAtlasSets, FontAtlasWarning, PositionedGlyph, Text, TextAlignment, TextSection,
-    TextSettings, YAxisOrientation,
-};
-use ab_glyph::PxScale;
-use bevy_asset::{AssetId, Assets, Handle};
-use bevy_ecs::component::Component;
-use bevy_ecs::prelude::ReflectComponent;
-use bevy_ecs::system::Resource;
+use std::sync::{Arc, Mutex};
+
+use bevy_asset::{AssetId, Assets};
+use bevy_ecs::{component::Component, reflect::ReflectComponent, system::Resource};
 use bevy_math::Vec2;
-use bevy_reflect::prelude::ReflectDefault;
-use bevy_reflect::Reflect;
+use bevy_reflect::{std_traits::ReflectDefault, Reflect};
 use bevy_render::texture::Image;
 use bevy_sprite::TextureAtlas;
-use bevy_utils::HashMap;
-use glyph_brush_layout::{FontId, GlyphPositioner, SectionGeometry, SectionText, ToSectionText};
+use bevy_utils::{tracing::warn, HashMap};
 
-#[derive(Default, Resource)]
-pub struct TextPipeline {
-    brush: GlyphBrush,
-    map_font_id: HashMap<AssetId<Font>, FontId>,
+use cosmic_text::{Attrs, Buffer, Metrics, Shaping, Wrap};
+
+use crate::{
+    error::TextError, BreakLineOn, Font, FontAtlasSets, FontAtlasWarning, PositionedGlyph,
+    TextAlignment, TextSection, TextSettings, YAxisOrientation,
+};
+
+// TODO: cache buffers / store buffers on the entity
+// TODO: reconstruct byte indices
+// TODO: rescale font sizes in all examples
+// TODO: fix any broken examples
+// TODO: solve spans with different font sizes, see https://github.com/pop-os/cosmic-text/issues/64
+// TODO: (future work) split text entities into section entities
+// TODO: (future work) text editing
+// TODO: font validation
+// TODO: the only reason we need a mutex is due to TextMeasure
```

Comment:
@nicoburns, you may have opinions here.

Reply:
Taffy 0.4 (specifically the measure function / context rework) should completely remove the need for a Mutex. We just need to get a release out. I would consider https://github.com/DioxusLabs/taffy/milestone/6 an accurate reflection of the work needed to do that (we don't necessarily need to block release on the documentation work, but it would certainly be nice to have).

The model with 0.4 will probably be something like: store the cosmic-text `Buffer` and potentially some kind of cache as "node context" on the Taffy node. Then borrow `FontSystem` in the measure function when calling `compute_layout`.